[14.0] OCA' Odoo instance docker image

petrus-v opened the pull request:
Here is ongoing work to build OCA' Odoo docker image.

As discussed w/ @sbidoul and the OCA Internal Working Group, here we give a try to include OCA instance project settings to build the images that may be deployed in test/production environment.

We can move that in an independent repository if we needs different maintainers or if it makes confusions ! We through starting here avoid extra repositories and would works as long oca-custom contains (I suppose) module that are only used by OCA ! Anyway I think existing CI/CD shouldn't be that much impacted by this exploration

## Tasks

- [x] create module oca_all with all installed modules
- [x] add pyproject.toml create developer python environment using `uv` and `hatch-odoo`
- [x] Setup Docker image based on https://github.com/acsone/odoo-bedrock 
- [x] Setup github action to build 
- [x] In the Dockerfile, add an entry point that runs click-odoo-update (always)


## Optionnal/To discuss

- [ ] ~In the Dockerfile, add an entry point that neutralizes the database if we are in a non-prod environment.~
* [ ] ~Get exacted module version in odoo database and freeze like in production (I'd like to give a try using latest module version but maybe I'm too reckless !) advices welcomes~ => let's try with fresh modules
* [x] Install database using the released image before tag it

petrus-v committed:
[ADD] oca_all: module with all OCA instance dependencies

petrus-v committed:
[IMP] OCA's Odoo: odoo-community.org development/release

Add contribute tools in order to develop and release
docker images for https://odoo-community.org using
Acsonne toolsets likes hatch-odoo, odoo-bedrock...

This helps to quickly create a locale environement close
to what is installed on our production instance.

Use following unreleased code:

* odoo14-addon-membership-delegated-partner-line from OCA/vertical-association/pull/151
* odoo from OCA/OCB/pull/1299

Co-authored-by: Stéphane Bidoul <stephane.bidoul@acsone.eu>
Co-authored-by: Holger Brunn <mail@hunki-enterprises.com>

sbidoul commented on repos.yaml:
```diff
@@ -0,0 +1,25 @@
+    - oca 14.0
+  shell_command_after:
+    # Support python -m odoo, for debugging configuration comfort.
+    - curl -sL https://github.com/odoo/odoo/pull/81864.patch | git am
```
We should backport this to OCB. I had started a while ago: https://github.com/OCA/OCB/pull/1132

petrus-v replied:
Thanks a lot, your [backport has been merged in v14.0](https://github.com/OCA/OCB/pull/1288) We can rely on OCB \o/

sbidoul commented on CONTRIBUTE.md:
````diff
@@ -0,0 +1,78 @@
+### Update a specific OCA module dependency using the latest pypi release
+
+```bash
+uv sync -P odoo14-addon-<module-name-underscores-replaced-by-hyphens>
````
```suggestion
uv sync --upgrade-package odoo14-addon-<module_name>
```

No need to replace underscores here?

petrus-v replied:
I'll give a try, I through it's the name of the package which is declared with hyphens in python project

sbidoul commented on pyproject.toml:
```diff
@@ -0,0 +1,153 @@
+    "PyGithub<2.0.0",
+]
+addons_dirs = ["./"]
+odoo_version_override = "14.0"
```
```suggestion
```

Not necessary?

sbidoul commented on pyproject.toml:
```diff
@@ -0,0 +1,153 @@
+odoo = { git = "https://github.com/petrus-v/odoo", branch = "14.0-oca-20250317" }
+odoo-addons-design-themes = { git = "https://github.com/petrus-v/design-themes", branch = "14.0-oca-20250317" }
+# PR: https://github.com/OCA/vertical-association/pull/151
+odoo14-addon-membership-delegated-partner-line = { git = "https://github.com/gdgellatly/vertical-association", branch = "gdg-membership_delegated_partner_line", subdirectory = "setup/membership_delegated_partner_line"}
```
```suggestion
odoo14-addon-membership-delegated-partner-line = { git = "https://github.com/OCA/vertical-association", ref = "refs/pull/151/head", subdirectory = "setup/membership_delegated_partner_line"}
```

petrus-v replied:
I've already test that, but `ref` doesn't seems to be supported by uv even on the current released moving uv from 0.5.11 to 0.68 version but doesn't seems any change (there is the new package field)

```
unknown field `ref`, expected one of `git`, `subdirectory`, `rev`, `tag`, `branch`, `url`, `path`, `editable`, `package`, `index`, `workspace`, `marker`, `extra`, `group`
```

but `rev` seems to work as expected \o/ !

sbidoul reviewed (commented):
Looks good!

sbidoul commented on Dockerfile:
```diff
@@ -0,0 +1,65 @@
+  && apt -y install --no-install-recommends \
+       git \
+       python3.9-dev \
+       libldap-dev \
+       libsasl2-dev \
```
```suggestion
```

We don't need python-ldap.

sbidoul commented on pyproject.toml:
```diff
@@ -0,0 +1,115 @@
+    "psutil==5.6.6",
+    "psycopg2==2.9.10",
+    "pydot==1.4.1",
+    "python-ldap==3.1.0",
```
```suggestion
```

sbidoul commented on docker-compose.yml:
```diff
@@ -0,0 +1,53 @@
+      - DB_USER=${POSTGRES_USER:-oca}
+      - DB_NAME=${POSTGRES_DB:-oca}
+      - UNACCENT=True
+      - WORKERS=2
```
```suggestion
      - WORKERS=0
```

Better for development?

petrus-v replied:
The intent of the docker compose is mainly to use in CI to at test installing oca_all in the docker container !

It wasn't implemented when you review, any way no needs multiples odoo workers to install odoo I've followed your suggestion.

petrus-v committed:
[FIX] website_oca_psc_team: fix tour

petrus-v commented on .github/workflows/build-oca-image.yml:
```diff
@@ -0,0 +1,73 @@
+          docker compose run --rm odoo odoo -c /etc/odoo.cfg -i oca_all --without-demo= --stop-after-init
+
+      -
+        name: Login to ghcr.io
+        if: ${{ github.repository_owner == 'OCA' && github.ref == 'refs/heads/14.0' || startsWith(github.ref, 'refs/tags/') }}
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
@sbidoul I suppose some github configuration  are require here if we want to push the docker image in the github registry !

sbidoul replied:
I think actions have write access to the registry by default.

petrus-v commented on .github/workflows/build-oca-image.yml:
```diff
@@ -0,0 +1,73 @@
+          cache-to: type=inline
+          push: true
+
+      -
+        name: Push image with tag if tagged
+        if: ${{ github.repository_owner == 'OCA' && startsWith(github.ref, 'refs/tags/') }}
+        run: |
+          docker tag ghcr.io/oca/oca-custom:14.0-oca-latest ghcr.io/oca/oca-custom:${{  github.ref_name }}
+          docker push ghcr.io/oca/oca-custom:${{  github.ref_name }}
```
@sbidoul I'm not sure how we are going to manage proper release, here if we add a git tag to this repo, this should tag the docker image with the same tag

sbidoul replied:
Or maybe we don't need to publish a docker image and we can build it on the deployment server.
We would still reach our goal of having a reproducible environment for developers.

petrus-v replied:
I don't really like building docker image while deploying, it's like I've no warranty that what I've tested on staging would be exactly the same in production server. Lets think loudly and enumerate risks:

* mainly non python dependencies with the uv stuffs we are pretty sure to get the same python environnement
* except if pyproject reference an open PR which I don't know how to prevent such practice to be merged (maybe we can do a ticket for later as an extra pre-commit hook and avoid this to be happen at review for the time being)
  
  
So I'm fine as long:
* pyrproject retreives only pypi released package
* I don't know about OCA servers environnement
* you feel confident doing that way
* this will make github action maintenance easier !

sbidoul replied:
> I don't really like building docker image while deploying

I usually don't do that either.

But with uv.lock, the risk of getting a different build is virtually zero. And other dependencies are the bedrock base image, itself based on ubuntu, which are both very stable. 

Even if we source open PRs, uv.lock has the commit in addition to the rev, so if that has changed, we'll know it, I think. Although I'm curious to see how uv behave if the PR has changed in the meantime.

So I tend to favor simplicity with one less thing to maintain (i.e. the image build pipeline), but I'm also fine to work with tagged images. As you prefer.

petrus-v replied:
if we push the latest to a docker registry from the 14.0 branch build we may benefit use it in CI as docker cache but can be done later if we needs it

petrus-v replied:
Sorry I've removed previous un-related message...

> Although I'm curious to see how uv behave if the PR has changed in the meantime.

This is my main concerns, if tests takes time on staging and we are not able to reproduce the same in production this would  add extra stress to administrators while deploying to production.

Also I'm wondering how to easily gives some feedback to developers/administrator to know which version is currently on which environnement staging/production (actually I don't know if there is any `staging` and test could happen on runboat in such case then the missing test would be migrating from one version to an other... :thinking: )

An other topic is the risk of inconsistance of the oca-custom repo and the pyproject, so the tests pipline wouldn't test what we are deplopying.

petrus-v commented on docker-compose.yml:
```diff
@@ -0,0 +1,53 @@
+      POSTGRES_DB: ${POSTGRES_DB:-oca}
+      POSTGRES_USER: ${POSTGRES_USER:-oca}
+      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD:-ocapwd}
+      # POSTGRES_HOST_AUTH_METHOD: "password"
```
to remove

petrus-v commented on repos.yaml:
```diff
@@ -0,0 +1,11 @@
+./src/design-themes:
```
should we create an OCA fork instead and remove this git aggregator configuration ?

sbidoul replied:
we could also add the `design-themes` repo as a submodule that we add in `tool.hatch-odoo.addons_dirs` in `pyproject.toml`.

sbidoul replied:
so we would not need git-aggregator at all

petrus-v replied:
By default I don't really like git submodules ! 

I get the point, no needs to make design-themes a python package, I've the feeling of an anti-pattern here :thinking: ! That's pragmatic solution :) ! 
I'm wondering how to prevent this list of git submodules to grow over the time.

In fact we could also get ride of gitaggregator configuration and point to the acsone branch: `acsone/14.0-with-pyproject`,  there are no much activity on this branch:

![image](https://github.com/user-attachments/assets/9a977c67-a03a-46c5-8744-cfc68cd376a3)

latest odoo commit date November 2024
latest odoo fix commit date April 2023

But I suppose others can complains this branch wouldn't be secure enough as not under OCA Umbrella... 

From my point of view using the `acsone/14.0-with-pyproject` branch is the easiest solution to me that we could improve later, or do you really prefer the git submodules solution ?

sbidoul replied:
> By default I don't really like git submodules !

Me neither. I'm just exploring alternatives because you raised the point :)

An OCB-like fork with automatic synchronization of design-themes also seems too much work.

> From my point of view using the acsone/14.0-with-pyproject branch is the easiest solution to me that we could improve later

I have just rebased it.

petrus-v commented on uv.lock:
```diff
@@ -0,0 +1,2094 @@
+
+[[package]]
+name = "odoo14-addon-github-connector"
+version = "14.0.2.3.1.dev1"
```
@sbidoul For all `oca_all` module dependencies, I initially used the latest versions available on PyPI when running `uv sync` for the first time. Would you like me to gently freeze them (only in `uv.lock`) to match the versions used in production, so that a plain `uv sync` ensures exact same dependencies, while `uv sync -U` allows updates?

sbidoul replied:
It's probably not worth the effort. I suspect they have not changed that much since they were last updated in prod. So let's try with the latest versions.

petrus-v reviewed (commented):
@sbidoul I think we are pretty closed to get it.

I'm not sure how to properly build the image that would contains un-merged dependencies, but it's probably better practice to wait to get it merged before deploy in production (I'll try to help there in vertical-association)!!

sbidoul commented:
In the Dockerfile, we could add an entry point that runs `click-odoo-update` (always) and neutralizes the database (if we are in a non-prod environment).

A potential starting point: https://github.com/acsone/acsone-odoo-project-template/blob/master/src/container/entrypoint-dbbase.jinja

sbidoul commented on CONTRIBUTE.md:
```diff
@@ -0,0 +1,78 @@
+
+## Concepts
+
+This repository is setup as other OCA's repositories to launch CI as usual and as an
+extra configuration in order to realease the OCA' Docker image used by our Odoo
+instance.
```
```suggestion
This repository is setup as other OCA's repositories to launch CI as usual and as an
extra configuration in order to build the OCA' Docker image used by our Odoo
instance, as well as facilitate the bootstrapping of a development environment.
```

sbidoul commented on CONTRIBUTE.md:
```diff
@@ -0,0 +1,78 @@
+- Postrgesql
+- [uv](https://docs.astral.sh/uv/)
+- Some dependencies to be able to build some python packages: `libpq-dev`,
+  `build-essential`, TODO
```
+ wkhtmltopdf

sbidoul commented on CONTRIBUTE.md:
````diff
@@ -0,0 +1,78 @@
+### Update OCB Branch
+
+```bash
+uvx --from git-aggregator gitaggregate -c repos.yaml -d ./src/ocb -p
````
```suggestion
```

sbidoul commented on CONTRIBUTE.md:
```diff
@@ -0,0 +1,78 @@
+- Some dependencies to be able to build some python packages: `libpq-dev`,
+  `build-essential`, TODO
+
+Prepare a python virtual environment and install the required dependencies:
```
```suggestion
Prepare a python virtual environment with the correct python version (which uv will download for you if necessary) and install the required dependencies:
```

petrus-v commented:
@sbidoul 

> and neutralizes the database if we are in a non-prod environment.

odoo neutralizes is availaible from version 16.0, do you mean using @hbrunn's PR https://github.com/OCA/server-backend/pull/346 ?

petrus-v commented on .github/workflows/build-oca-image.yml:
```diff
@@ -0,0 +1,37 @@
+        run: |
+          docker compose up --wait db kwkhtmltopdf
+          docker compose run --rm odoo odoo -c /etc/odoo.cfg -i oca_all --without-demo= --stop-after-init
+          # purpose it to test update entrypoints
+          docker compose up --wait odoo
```
This is not strictly necessary and takes some time likes `-u all`. This test some how 000_10_odoo-update entry point with a database installed  (it not strictly assert that the update has been done). While the -i oca_all test without odoo database!

Any opinion on this?

sbidoul commented:
> @sbidoul
> 
> > and neutralizes the database if we are in a non-prod environment.
> 
> odoo neutralizes is availaible from version 16.0, do you mean using @hbrunn's PR [OCA/server-backend#346](https://github.com/OCA/server-backend/pull/346) ?

I meant a minimal custom script to disable the crons and mail servers.

Anonymization can come on top but I'd apply that manually?

petrus-v commented:
> @sbidoul
> 
> > and neutralizes the database if we are in a non-prod environment.
> 
> odoo neutralizes is availaible from version 16.0, do you mean using @hbrunn's PR [OCA/server-backend#346](https://github.com/OCA/server-backend/pull/346) ?

In fact hbrunn's works is about anonymization, not neutralization. So I'm going to start a dedicated script to at least deactivate emails and crons.

And I believes that we should keep neutralisation as sql (likes in later odoo version) in order to execute them before update!

So at the end we will go with: neutralization (if not prod) > update > anonymization (if not prod)

petrus-v commented:
I've missed your comment you made before mine, Sorry I'm feeling noisy,
sometimes I think loudly...

This commit is a proposition to neutralize the database
https://github.com/OCA/oca-custom/pull/59/commits/209f66523f6414a61afa4fd626100ef3a3796fe7

Those SQL script cames and are adapted from Odoo 18.

I'm wondering if there are some thing to do with github connector, as far I
can read github token is in config file so I guess it's safe as long
production token wouldn't be in other environements

Le lun. 31 mars 2025 à 09:30, Stéphane Bidoul ***@***.***> a
écrit :

> @sbidoul <https://github.com/sbidoul>
>
> and neutralizes the database if we are in a non-prod environment.
>
> odoo neutralizes is availaible from version 16.0, do you mean using
> @hbrunn <https://github.com/hbrunn>'s PR OCA/server-backend#346
> <https://github.com/OCA/server-backend/pull/346> ?
>
> I meant a minimal custom script to disable the crons and mail servers.
>
> Anonymization can come on top but I'd apply that manually?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/oca-custom/pull/59#issuecomment-2765358157>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBAYOZI3R2VDUQRHFEF5632XDVILAVCNFSM6AAAAABZF4VPW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVGM2TQMJVG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: sbidoul]*sbidoul* left a comment (OCA/oca-custom#59)
> <https://github.com/OCA/oca-custom/pull/59#issuecomment-2765358157>
>
> @sbidoul <https://github.com/sbidoul>
>
> and neutralizes the database if we are in a non-prod environment.
>
> odoo neutralizes is availaible from version 16.0, do you mean using
> @hbrunn <https://github.com/hbrunn>'s PR OCA/server-backend#346
> <https://github.com/OCA/server-backend/pull/346> ?
>
> I meant a minimal custom script to disable the crons and mail servers.
>
> Anonymization can come on top but I'd apply that manually?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/oca-custom/pull/59#issuecomment-2765358157>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBAYOZI3R2VDUQRHFEF5632XDVILAVCNFSM6AAAAABZF4VPW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVGM2TQMJVG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Pierre

petrus-v commented:
I'm still wondering what would be the best approach for building and deploying Docker images. One concern I have is ensuring reliability when building for production. To mitigate risks, I’d like to implement a system that stores tagged Docker images explicitly.

### Tagging Strategy Proposal

I’m considering using a set of moving tags, something like:

  *  prod, old-prod
  *  staging, old-staging

This would allow us to safely promote builds through staging to production while keeping a rollback path available. Here's an example based on the following Git history and tag structure:

```gitattributes
* a1d5140 
* 7351e07
* 5febd12 (tag: staging)
* f1144e6 
*   126e946 Merge PR #58 into 14.0
|\  
| * 84f82f7 
| * ...
| * 0b35543
|/  
* 8413442
* f3bf40a
* d024d6d (tag: old-staging, tag: prod)
* 0eabfd6
* 2e2e92a
*   cde549e Merge PR #55 into 14.0
|\  
| * 00aec5c 
| * ... 
| * d74d661 
|/  
* d8bdea5
* 99fdf6e (tag: old-prod)
*   e06edb4 Merge PR #54 into 14.0
```

### Deploying to Production

The deployment process could follow these steps:

* Update Git tags:
  * Move `old-prod` to the current `prod` tag reference.
  * Move `prod` to point to the current `staging` tag reference.
* Trigger a GitHub workflow when the prod tag changes (or from the same script):
  * Pull the current `prod` docker image and push it as `old-prod`.
  * Pull the current `staging` image and push it as `prod`.
* Deploy the `prod` image either from within the CI or after a successful CI run.

After deployment, the Git history would look like this:

```gitattributes
* a1d5140 
* 7351e07
* 5febd12 (tag: staging, tag: prod)
* f1144e6 
*   126e946 Merge PR #58 into 14.0
|\  
| * 84f82f7 
| * ...
| * 0b35543
|/  
* 8413442
* f3bf40a
* d024d6d (tag: old-staging, tag: old-prod)
* 0eabfd6
* 2e2e92a
*   cde549e Merge PR #55 into 14.0
|\  
| * 00aec5c 
| * ... 
| * d74d661 
|/  
* d8bdea5
* 99fdf6e
*   e06edb4 Merge PR #54 into 14.0
```

In case of a production issue, we can quickly roll back to the old-prod image.

To further decouple staging from prod, we could also store Docker images using the commit hash directly

### Staging Deployment

A staging deployment could follow this logic:

* Update Git tags:
  * Move `old-staging` to the current `staging` tag.
  * Point `staging` to a new commit reference.
* Trigger a GitHub workflow when the `staging` tag changes (or from the same script):
  * Pull the current `staging` image and tag it as `old-staging`.
  * Build (or re-tag an existing image by commit hash) and tag it as `staging`.
* Deploy the `staging` image after a successful CI run.

In the case of staging, maintaining `old-staging` might not be critical, simplifying the process.

### Benefits of This Approach

* Provides explicit visibility for developers on what's currently deployed in production.
* same docker image is used from staging to production
* Ensures that the previous, current and future production images are preserved in the Docker registry.
* Makes rollbacks fast and safe.

sbidoul commented:
This is looking good. Thanks a lot for this work!

About the tagging strategy, my usual approach is to create version-like git tags and corresponding image tags. To deploy, it is then just a matter of updating the tag in docker-compose.yml or some associated .env file. And when a version is deemed good for production, do the same update in prod.

That said, I have no real preference on that strategy for the OCA instance, so your proposal is good to me too.

petrus-v commented:
I've changed a bit my mind because in my previous messages I was trying to manage 3 concerns:

* **Provides explicit visibility for developers on what's currently deployed in production.*
* **create release workflow**
* **document and script the deployment**

So I end up with a solution that I've tested on my fork that I feel elegant enough:
* I've discover we can trigger manual workflow from actions panel in github, so I've setup a workflow to bump oca-custom package version (using bump-my-version), assuming we configure a Personnal Access Token https://github.com/OCA-git-bot, ony once could bump the version and push a tag on the HEAD of the selected branch. 
* once a tag is pushed CI will build docker image with the same tag

> **Note**: Of course someone allowed to push tag can do it directly a released image will be build as well

To provide explicit visibility for developers on what's currently deployed in production, I'm thinking a way to add environement variable in the docker image and display such value somewhere accessible in odoo.

For deployment, as you suggest, changing version of the docker image in a docker compose file (or .env) and up the project will do the job.

The last missing things here (from my point of view) is the unreleased dependency where I give some helps: https://github.com/OCA/vertical-association/pull/151 which is technically ready to merge but probably needs some attention to make sure what is on the production server !

sbidoul commented:
@petrus-v regarding unreleased dependencies, it's probably reasonable to remove this CI check from this repo.

Regarding the release process, why not creating the tags manually? PAT are a kind of security risk and a bit of pain to manage since they expire, so if we can avoid them I tend to prefer.

petrus-v commented:
> regarding unreleased dependencies, it's probably reasonable to remove this CI check from this repo.

hum... I think we are enough in the group to do reviews and merge PRs before deployment, If needs I'm ready to help on a series of modules/repos to maintains. If we are in the hurry with current workflow we can even push a tag on any commit (even it's not part of the 14.0 branch it should create the docker image.

> Regarding the release process, why not creating the tags manually? PAT are a kind of security risk and a bit of pain to manage since they expire, so if we can avoid them I tend to prefer.

I wanted to change the version in the pyproject.toml which require to push commit on 14.0 branch, this is not strictly required. Also I was wondering who are able to push tag in the repository ?

While it's not recommend it's possible to create PAT without expiration !

I guess people that get access to push tag won't get access push access to the 14.0 branch so I can get ride of the manual workflow and the bump-my-version configuration from this PR?

hbrunn commented on CONTRIBUTE.md:
````diff
@@ -0,0 +1,98 @@
+
+```bash
+uv sync -U
+```
````
I must admit I've never heard of `uv` before looking at this PR, so probably dumb question: What's the recommended way of editing code here?

Now I'm toying with this to get familiar with the process, I do

```sh
git clone $repo -b 14.0
cd $repo
pip install -e .
# do my stuff
# later
git commit -m $message && git push ...
```

is that's how it's meant to go?

sbidoul replied:
Hi Holger, you probably missed the *Setup developer environment* section that this PR adds to the README.
Please have a look and if that is still unclear we should improve that section.

hbrunn replied:
that's what I'm commenting on. It tells me to run `uv sync`, which I did. Now I've all the modules available in the project's python path, that's great, but what's the recommended way to actually do development now?

sbidoul replied:
`uv sync` has created a virtualenv named `.venv` in the project directory.

The project itself is installed in editable mode and the addons are in addons_path, so you can simply edit the code, and run odoo with `uv run odoo` as mentioned above. This is more or less equivalent to running `.venv/bin/python -m odoo`, so you can configure your favorite IDE/editor/debugger with that.

If you add dependencies, just run `uv sync` again. `uv run odoo` will also re-sync automatically.

`uv` does not install `pip`, but you can run `uv pip list` to inspect what is installed in the venv.

hbrunn replied:
thanks, so far it's clear. But I'm talking about developing stuff for addons in other repos, which will probably mostly happen - those are just in the python path to my understanding?

Let's say OCA asks for some patch to mis_builder, what do I do?

sbidoul replied:
You add it to the [tool.uv.sources](https://docs.astral.sh/uv/reference/settings/#sources) section of `pyproject.toml` and `uv sync` / `uv run` again. There are a couple already with git references to branches and PRs, but I think you can add a local directory too.

So it would likely be something like

```toml
[tool.uv.sources]
odoo14-addon-mis_builder = { path = "/home/.../OCA/mis-builder/setup/mis_builder", editable = true }
```

which you then replace with a git reference when you have done the mis-builder PR.

sbidoul replied:
And of course, odoo `--addons-path` continues to work normally too.

hbrunn replied:
great, thanks. will play a bit with this and propose a section about that above

hbrunn commented on CONTRIBUTE.md:
```diff
@@ -0,0 +1,98 @@
+
+Requirements:
+
+- Postrgesql
```
```suggestion
- Postgresql
```

hbrunn commented on docker-compose.yml:
```diff
@@ -0,0 +1,67 @@
+      # install is done without checksum so start in CI will be quite long
+      interval: 2s
+      timeout: 200ms
+      retries: 150
```
```suggestion
      retries: 150
    depends_on:
      - db
      - kwkhtmltopdf
```

hbrunn commented on CONTRIBUTE.md:
````diff
@@ -0,0 +1,98 @@
+```bash
+uv sync
+```
+
````
```suggestion

### Development

For addons living in this repository, you can just change code and restart Odoo with the `uv run` command.

For addons in other repositories, the procedure is as follows:

- check out the repository somewhere, ie /src/$repo
- add the following line to ``pyproject.toml`` in the `[tool.uv.sources]` section:

        odoo14-addon-$youraddon = { path = "/srv/$repo/setup/$youraddon", editable = true }
     
- run `uv sync`
- restart Odoo
```

petrus-v commented:
> > regarding unreleased dependencies, it's probably reasonable to remove this CI check from this repo.
> 
> hum... I think we are enough in the group to do reviews and merge PRs before deployment, If needs I'm ready to help on a series of modules/repos to maintains. If we are in the hurry with current workflow we can even push a tag on any commit (even it's not part of the 14.0 branch it should create the docker image.
> 
> > Regarding the release process, why not creating the tags manually? PAT are a kind of security risk and a bit of pain to manage since they expire, so if we can avoid them I tend to prefer.
> 
> I wanted to change the version in the pyproject.toml which require to push commit on 14.0 branch, this is not strictly required. Also I was wondering who are able to push tag in the repository ?
> 
> While it's not recommend it's possible to create PAT without expiration !
> 
> I guess people that get access to push tag won't get access push access to the 14.0 branch so I can get ride of the manual workflow and the bump-my-version configuration from this PR?

I've removed the manual step that would have require PAT (Personal Access Token), bump my version configuration and update the release documentation by this commit https://github.com/OCA/oca-custom/pull/59/commits/79512d328af1351fb119b80fa79afb0265747b54

petrus-v committed:
[CI] ignore unreleased dependencies

petrus-v commented:
@sbidoul 

* This commit https://github.com/OCA/oca-custom/pull/59/commits/1b2ef6a73e452be36539207ed079c9c22e347997 ignore un-released dependencies as you suggested earlier, but I'm wondering the impacted on the released process, notably if we add dependency to a new module that won't be present on pypi - it's the case here-, as far I remember we can't publish a module on pypy package that dependents on something that is not released?

* This commit https://github.com/OCA/oca-custom/pull/59/commits/d8ae21df89cec3d870af273d92574067a151de65  use the PR https://github.com/OCA/OCB/pull/1299, I haven't deep into to make sure runboat do not use the PR on OCA/OCB repo as odoo code source [(cf comment)](https://github.com/OCA/OCB/pull/1299#issuecomment-2935648779) but could be nice to go ahead on this topic add a tag and release a version to deploy on a staging environment that Virginie and Rebacca could tests what we do. An Image is ready [on my fork](https://github.com/petrus-v/oca-custom/pkgs/container/oca-custom) `docker pull ghcr.io/petrus-v/oca-custom:v14.0.20250604`. The Idea would be to merge here and add a tag in this repo to run the release pipeline and publish the docker image here !

* This commit https://github.com/OCA/oca-custom/pull/59/commits/07c368083cf220fcbea26da1daab6a36c4284d27 revert using OCB's PR in this repo CI as it makes runboat error, but keep it in the released image : this makes a difference between what's tested and what's released, I'm starting to considering to create a different repository or improve the tool chain (that could be done later if we really have a dependency in unittest or want test happens on runboat). Also those issue could be mitigate by finding a way to improve reviews in the workgroup (I'll probably suggest ideas on this topic by email).

* while doing this I've open a proposal on odoo-bedrock here: https://github.com/acsone/odoo-bedrock/pull/53 [seen in this build](https://github.com/OCA/oca-custom/actions/runs/15441653686/job/43460668488?pr=59) ;) I've to investigate a bit more why CI sudently failled to see postgresql service which was not the case on this commit https://github.com/OCA/oca-custom/pull/59/commits/d8ae21df89cec3d870af273d92574067a151de65... :exploding_head: 

Any comment welcome :)

ps: If I can help with admin sys tasks I could help if needs.

sbidoul commented:
> impacted on the released process

There might be something in the bot to prevent the publishing of some addons, or the whole repo. I'll need to check. 
At least `oca_all` does not make sense for publishing.

I'll find a moment to deploy this on test.odoo-community.org.

petrus-v commented:
I"ve squashed some commits to get a cleaner history !

sbidoul committed:
Use standard mechanism to declare dev dependencies

sbidoul committed:
Use explicit docker cache id

Since some python wheels depend on the distro
on which they have been build (eg psycopg2), we use
an explicit cache id which includes the distro name.

sbidoul committed:
Add openupgradelib

It's necessary to run, e.g. github_connector migrations

sbidoul committed:
Tweak comment in entrypoint

sbidoul commented:
@petrus-v I sent https://github.com/petrus-v/oca-custom/pull/1 with a few minor tweaks, and deployed on https://test.odoo-community.org.

The docker compose currently builds from a local clone of this branch.

All looks good. Thanks for this work!

petrus-v committed:
Merge pull request #1 from acsone/14.0-tweak-build-sbi

Tweak build

petrus-v committed:
FIX oca_all dependencies

* account_statement_import_online_wise were missing on staging only

module missing from production 20250617

* l10n_ch_pain_base
* l10n_ch_pain_credit_transfer
* project_task_add_very_high
* web_search_with_and

and bumps all module versions

petrus-v committed:
freeze pytest version

waiting for https://github.com/camptocamp/pytest-odoo/pull/85

petrus-v committed:
[IMP] docker entrypoint install oca_all if empty database

petrus-v committed:
Add missing depenencies likes in production

petrus-v committed:
kwkhtmltopdf use init:true to quickly stop properly propagate signals

petrus-v committed:
add click-odoo fix membership line maintenance script

sbidoul commented:
@petrus-v shall we merge this?

petrus-v commented:
@sbidoul I think the latest point without responses here is regarding build and publish process with some unreleased dependencies that may crash the publish CI pipeline that you know more than me.

Otherwise this seems working on test environment so I'm in favour to merge it !

petrus-v committed:
integrate account_statement_import_online_wise Tom's fix using foreing currency
